cpp/integ/micro_benchmarks.cc

```diff
@@ -0,0 +1,128 @@
+// Copyright (C) 2017-2018 Dremio Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdlib.h>
+#include <gtest/gtest.h>
+#include "arrow/memory_pool.h"
+#include "integ/test_util.h"
+#include "gandiva/projector.h"
+#include "gandiva/status.h"
+#include "gandiva/tree_expr_builder.h"
+
+namespace gandiva {
+
+using arrow::int32;
+using arrow::int64;
+using arrow::boolean;
+
+class TestBenchmarks : public ::testing::Test {
+ public:
+  void SetUp() { pool_ = arrow::default_memory_pool(); }
+
+ protected:
+  arrow::MemoryPool* pool_;
+};
+
+int32_t GenRandLessThan250(int field_num) {
+  auto num = rand();
+  return (num % 250);
+}
+
+int64_t GenRandInt64(int field_num) {
+  return rand();
+}
+
+TEST_F(TestBenchmarks, TimedTestAdd3) {
+  // schema for input fields
+  auto field0 = field("f0", int64());
+  auto field1 = field("f1", int64());
+  auto field2 = field("f2", int64());
+  auto schema = arrow::schema({field0, field1, field2});
+
+  // output field
+  auto field_sum = field("add", int64());
+
+  // Build expression
+  auto part_sum = TreeExprBuilder::MakeFunction("add",
+    {TreeExprBuilder::MakeField(field1), TreeExprBuilder::MakeField(field2)},
+    int64());
+  auto sum = TreeExprBuilder::MakeFunction("add",
+    {TreeExprBuilder::MakeField(field0), part_sum},
+    int64());
+
+  auto sum_expr = TreeExprBuilder::MakeExpression(sum, field_sum);
+
+  std::shared_ptr<Projector> projector;
+  Status status = Projector::Make(schema, {sum_expr}, pool_, &projector);
+  EXPECT_TRUE(status.ok());
+
+  int64_t elapsed_millis;
+  status = TimedEvaluate<arrow::Int64Type, int64_t>(schema,
+                                                    projector,
+                                                    GenRandInt64,
+                                                    100 * MILLION,
+                                                    16 * THOUSAND,
+                                                    elapsed_millis);
+  ASSERT_TRUE(status.ok());
+  std::cout << "Time taken for Add3 " << elapsed_millis << " ms\n";
+}
+
+TEST_F(TestBenchmarks, TimedTestBigNested) {
```

Comment:
You could probably use `gbenchmark` instead of `gtest` for this. For an example usage see https://github.com/apache/arrow/blob/master/cpp/src/arrow/builder-benchmark.cc

Reply:
TimedEvaluate measures the time for a portion of the invocation; specifically the evaluate() method. The time taken to perform data generation is not measured.

Does gbenchmark handle this?